### ISSUE TITLE
DataBackendRaster: Extraction many non-continuous cell values is very slow

### ISSUE THREAD

**Issue by pat-s** [mlr-org/mlr3spatial#26]
The call never finishes due to the slowness and should actually finish in less than 10s.
The reason is that the single cell extraction is quite slow.

```r
library("mlr3")
library("mlr3tuning")
library("mlr3spatial")
library("paradox")

logger = lgr::get_logger("bbotk")
logger$set_threshold("debug")

logger_mlr3 = lgr::get_logger("mlr3")
logger_mlr3$set_threshold("debug")

logger_mlr3tuning = lgr::get_logger("mlr3tuning")
logger_mlr3tuning$set_threshold("debug")

learner = lrn("classif.rpart")
tune_ps = ps(
  cp = p_dbl(lower = 0.001, upper = 0.1),
  minsplit = p_int(lower = 1, upper = 10)
)
terminator = trm("evals", n_evals = 2)
tuner = tnr("grid_search")
stack_classif = demo_stack_spatraster(1)
value = data.table::data.table(ID = c(0, 1), y = c("negative", "positive"))
terra::setCats(stack_classif, layer = "y", value = value)
backend_sp = as_data_backend(stack_classif)
task = as_task_classif(backend_sp, target = "y", positive = "positive")

hout = rsmp("holdout")
measure = msr("classif.ce")

instance = TuningInstanceSingleCrit$new(
  task = task,
  learner = learner,
  resampling = hout,
  measure = measure,
  search_space = tune_ps,
  terminator = terminator
)
instance

tuner$optimize(instance)
```

**Comment by berndbischl** [comment]
can you please post a description of the problem and at least partial logging out put

**Comment by berndbischl** [comment]
also: are you reasonably sure that  the AutoTuner hangs? in which case you filed the issue at the wrong repo, should be mlr3tuning then.
if its mlr3spatial its your own issue :) and we can help much less

**Comment by berndbischl** [comment]
also as quick side note: mlr3verse is not really best for a reprex. due to it i am currently installing 50 packages, where 5 might have been enough. its not really "minimal" but maximal....

**Comment by berndbischl** [comment]
next observation - as expected: 
this has nothing to do with the AT, normal tuning produces the same error, this should always be checked and minimized....

**Comment by berndbischl** [comment]
also, this line
DataBackendRaster.R:
mlr3misc::calculate_hash(self$compact_seq, self$stack)

uses self$compact_seq, which does not even exist in the class definition...?

**Comment by berndbischl** [comment]
also in DataBackendRaster you use in the hash calc fun "compact_seq".
that object / member does NOT even exist in the class!

in seems to exist in DataBackendVecor, but you are not inheriting from it. 
this seems like a very bad bug? and i am not sure why you cannot see this in a test - before you file this here

in any case, that means that i cannot look further as the src pack is very much broken due to this

**Comment by berndbischl** [comment]
and.................... the bug has nothing to with mlr3tuning. i just ran the bechmark function on your generated design points manually. it still hangs.

this did take me about an hour.....

**Comment by berndbischl** [comment]
i mean, why not run this before pinging

```
learner = lrn("classif.rpart")
tune_ps = ParamSet$new(list(
  ParamDbl$new("cp", lower = 0.001, upper = 0.1),
  ParamInt$new("minsplit", lower = 1, upper = 10)
))
stack_classif = demo_stack_spatraster(1)
value = data.table::data.table(ID = c(0, 1), y = c("negative", "positive"))
terra::setCats(stack_classif, layer = "y", value = value)
backend_sp = as_data_backend(stack_classif)
task = as_task_classif(backend_sp, target = "y", positive = "positive")
resampling = rsmp("holdout")
g = benchmark_grid(tasks = task, learner = learner, resamplings = resampling)
benchmark(g)
```

**Comment by pat-s** [comment]
> also: are you reasonably sure that the AutoTuner hangs? in which case you filed the issue at the wrong repo, should be mlr3tuning then.

I opened it here because it only happens with the Backends of this package.

> also in DataBackendRaster you use in the hash calc fun "compact_seq".
that object / member does NOT even exist in the class!

Thanks for noticing, will clean that up.

> in seems to exist in DataBackendVecor, but you are not inheriting from it.

Both are on the same level and should not inherit from each other.

> i mean, why not run this before pinging

This runs just fine for me - do you get an error there?

Right now I still did not know what the problem is. Will try to check in more detail on a simple tuning scenario.

**Comment by berndbischl** [comment]
ok this is weird. the train-call does work for me now, maybe i was just tired or impatient yesterday night... sorry.
have you also tried the related benchmark call?

**Comment by berndbischl** [comment]
learner = lrn("classif.rpart")
stack_classif = demo_stack_spatraster(1)
value = data.table::data.table(ID = c(0, 1), y = c("negative", "positive"))
terra::setCats(stack_classif, layer = "y", value = value)
backend_sp = as_data_backend(stack_classif)
task = as_task_classif(backend_sp, target = "y", positive = "positive")
resampling = rsmp("holdout")
g = benchmark_grid(tasks = task, learner = learner, resamplings = resampling)
benchmark(g)

does this run or hang for you?

**Comment by berndbischl** [comment]
i am not sure how fast this call should terminate but this does seem to hang for me

**Comment by pat-s** [comment]
I tracked it down and found a solution.

The problem was the `else` block in 

https://github.com/mlr-org/mlr3spatial/blob/632d47ed0de3ec642866eb39f1076d8be837395b/R/DataBackendRaster.R#L94-L109

where we read individual cells that are not connected to each other.
This is way slower than reading in block mode. So essentially the call from the OP would finish at some point in time but is just very slow.

I've now modified it in a way that we always do a block read of the required rows and then do a subsetting which is conditional on the fact if we ask for a continuous sequence (e.g. `1:10`) or single cells (e.g. `c(1, 3, 5, 6)`.

Reading is fast now and all other consistency tests if we read the correct cells are also passing.

FYI @be-marc 

Thanks for your help @berndbischl, helped me to narrow down the issue!
